pkg/http/server.go

```diff
@@ -0,0 +1,184 @@
+	errorLogFunc := srv.ErrorLogFunc
+
+	// Create new HTTP listener.
+	var listener *httpListener
+	listener, err = newHTTPListener(
+		addrs,
+		tlsConfig,
+		tcpKeepAliveTimeout,
+		readTimeout,
+		writeTimeout,
+		updateBytesReadFunc,
+		updateBytesWrittenFunc,
+		errorLogFunc,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Wrap given handler to do additional
+	// * return 503 (service unavailable) if the server in shutdown.
+	wrappedHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&srv.requestCount, 1)
+		defer atomic.AddInt32(&srv.requestCount, -1)
+
+		// If server is in shutdown, return 503 (service unavailable)
+		srv.inShutdownMutex.Lock()
+		inShutdown := srv.inShutdown
+		srv.inShutdownMutex.Unlock()
+		if inShutdown {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			return
+		}
+
+		// Handle request using passed handler.
+		handler.ServeHTTP(w, r)
+	})
+
+	srv.listenerMutex.Lock()
+	srv.Handler = wrappedHandler
+	srv.listener = listener
+	srv.listenerMutex.Unlock()
+
+	// Start servicing with listener.
+	return srv.Server.Serve(listener)
+}
+
+// Shutdown - shuts down HTTP server.
+func (srv *Server) Shutdown() error {
+	srv.inShutdownMutex.Lock()
+	if !srv.inShutdown {
+		// Indicate the server is in shutdown.
+		srv.inShutdown = true
+		srv.inShutdownMutex.Unlock()
+
+		// Close underneath HTTP listener.
+		srv.listenerMutex.Lock()
+		err := srv.listener.Close()
+		srv.listenerMutex.Unlock()
+
+		// Wait for opened connection to be closed up to Shutdown timeout.
+		shutdownTimeout := srv.ShutdownTimeout
+		shutdownTimer := time.NewTimer(shutdownTimeout)
+		ticker := time.NewTicker(serverShutdownPoll)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-shutdownTimer.C:
+				return errors.New("timed out. some connections are still active. doing abnormal shutdown")
+			case <-ticker.C:
+				if atomic.LoadInt32(&srv.requestCount) <= 0 {
+					return err
+				}
+			}
+		}
+	}
+	srv.inShutdownMutex.Unlock()
+
+	// shutdown in progress
+	return errors.New("http server already in shutdown")
+}
+
+// NewServer - creates new HTTP server using given arguments.
+func NewServer(addrs []string, handler http.Handler, certificate *tls.Certificate) *Server {
+	var tlsConfig *tls.Config
+	if certificate != nil {
+		tlsConfig = &tls.Config{
+			PreferServerCipherSuites: true,
+			MinVersion:               tls.VersionTLS12,
+			NextProtos:               []string{"http/1.1", "h2"},
+		}
+		tlsConfig.Certificates = append(tlsConfig.Certificates, *certificate)
+	}
+
+	httpServer := &Server{
+		Addrs:               addrs,
+		ShutdownTimeout:     DefaultShutdownTimeout,
+		TCPKeepAliveTimeout: DefaultTCPKeepAliveTimeout,
+		listenerMutex:       &sync.Mutex{},
+		inShutdownMutex:     &sync.Mutex{},
+	}
+	httpServer.Handler = handler
+	httpServer.TLSConfig = tlsConfig
+	httpServer.ReadTimeout = DefaultReadTimeout
+	httpServer.WriteTimeout = DefaultWriteTimeout
```

Comment:
We don't need to set this value it is not useful and would cause frequent timeouts for slow clients. 30 secs is still a small number for situations minio is deployed.